# Ex. 2.1

First we need to assume that the norm here is Euclidean norm $||\cdot||_2$, we also believe this is the case for any norm appearing in this book. The statement in this problem is true under Euclidean norm, but not necessarily true for other norms. Let's consider two vectors in $v_1, v_2 \in \mathbb{R}^2 $, where $ v_1 = [4,4]^T, v_2 = [6,1]^T $. We can see that:
\begin{align*}
  ||v_1||_1 >& \,||v_2||_1 \\ 
  ||v_1||_2 <& \,||v_2||_2 .
\end{align*}

This indicates that for a given point, the nearest neighbor is not invariant with different choices of norms.
  
Let's come back to prove the statment in this question:
$$\mathrm{argmin}_k\left(\left\Vert t_k - \hat{y}\right\Vert_2\right) = \mathrm{argmax}_k\left(\hat{y}_k\right)$$
**Proof**
\begin{align*}
  \mathrm{argmin}_k\left(\left\Vert t_k - \hat{y}\right\Vert_2\right) 
  =&\mathrm{argmin}_k\left(\left\Vert t_k - \hat{y}\right\Vert_2^2\right)\\
  =&\mathrm{argmin}_k\sum_{i=1}^K\left( t_{k\,i} - \hat{y}_i\right)^2 \\
  =&\mathrm{argmin}_k\sum_{i=1}^K\left(t_{k\,i}^2 + \hat{y}_i^2 - 2t_{k\,i}\hat{y}_i\right)\\
  =&\mathrm{argmin}_k\left(1+\left\Vert\hat{y}\right\Vert_2^2-2\sum_{i=1}^Kt_{k\,i}\hat{y}_i\right)\\
  =&\mathrm{argmin}_k\left(1+\left\Vert\hat{y}\right\Vert_2^2-2\hat{y}_k\right) \\
  =&\mathrm{argmin}_k\left(-2\hat{y}_k\right) \\
  =&\mathrm{argmax}_k\left(\hat{y}_k\right)
\end{align*}∎

# Ex. 2.2

The Bayes decision boundary is atainable because we know how the two classes (Blue and Organge) are generated. Let's first review on how these two classes are generated. We define a Gaussian bivariate distribution for each of the two classes, with the covariance matrix being the same $2\times2$ identity matrix $\mathbf{I}$, and mean vectors different. For the Blue class, the mean vector is $[1,0]^T$, and for the Organge $[0,1]^T$. Namely $\mathbf{Z}_1 = N\left([1,0]^T, \mathbf{I}\right), \mathbf{Z}_2 = N\left([0,1]^T, \mathbf{I}\right)$. For each of the two classes, we sample 10 times from its corresponding bivariate Gaussian vector. We denote the two set of samples as: $\mathbf{p}=\{p_i\}_{i=1}^{10}$ for the Blue and $\mathbf{q}=\{q_i\}_{i=1}^{10}$ for the Orange. Next, for each of the two classes, we first uniformly pick an element form $\mathbf{p}$ (or $\mathbf{q}$) to be the mean vector, and use $\frac15\mathbf{I}$ as the covariance matrix to form another bivariate Gaussian vector to generate a single realization. Such process is repeated for 100 times for each class.

Now let's first look for the Bayes decision boundary conditional on $\mathbf{p}$ and $\mathbf{q}$, $\Gamma(\mathbf{p}, \mathbf{q})\in \mathbb{R}^2$. For any $x\in\mathbb{R}^2$. The probability of $x$ being blue is:
$$P(x\,\mathrm{being\,blue})= \sum_{i=1}^{10}\frac1{10}\phi(x,p_i), $$
where $\phi(x,p_i)$ is the probability density function at $x$ of a bivariate Guassian with mean $p_i$ and covariance matrix $\frac15\mathbf{I}$. Then the Bayes decision boundary conditional on $\mathbf{p}$ and $\mathbf{q}$ should be those points where being blue and orange are equally likely:
$$\Gamma(\mathbf{p}, \mathbf{q})
=\{x\in\mathbb{R}^2\vert \sum_{i=1}^{10}\phi(x, p_i)=\sum_{i=1}^{10}\phi(x, q_i)\}$$

The final Bayes decision boundary should only be conditinal on given paramters $\mathbf{Z}_1$ and $\mathbf{Z}_2$, which should be the expectation of $\Gamma(\mathbf{p}, \mathbf{q})$ over all possible $\mathbf{p}$ or $\mathbf{q}$:
\begin{align*}
\Gamma(\mathbf{Z}_1, \mathbf{Z}_2) =& \Gamma(\mathrm{E}[\mathbf{p}],\mathrm{E}[\mathbf{q}])\\
=&\{x \in \mathbb{R}^2\,\vert\,\phi(x,[1,0]^T)=\phi(x,[0,1]^T)\}\\
=&\{x \in \mathbb{R}^2\,\vert\, x_1 = x_2\}.
\end{align*}

# Ex. 2.3

This excercise is related to ordered statistics, where the simplest case minimum is encountered. The volume of a $p$ dimmensional ball is proportional to $r^p$, where $r$ is its radius. Let's denote the random variable of the distance of any sampled point in this $p$ dimmentional unit ball to the origin as $Z$. Then,
\begin{align*}
P(Z < x)=&x^p, \\
P(Z > x)=&1-x^p, \\
P(Z_{(1)} > x) =& P(\prod_{i=1}^N Z_i > x)=\left(1-x^p\right)^N,\\
P(Z_{(1)} < x) =& 1 - \left(1-x^p\right)^N.
\end{align*}
The median is the $\frac12$ quantile of $Z_{(1)}$, solve for $1-\left(1-x^p\right)^N=\frac12$ gives us:
$$x=\left(1-\frac12^{1/N}\right)^{1/p}$$

# Ex. 2.4

$z_i$ is a linear combination of all the elements of any realization of $\mathbf{X}$, so it must be one dimensional normal. $\mathbf{X} \sim N(0, \mathbf{I})$, therefore we have:
\begin{align*}
\mathrm{E}[z_i]=& 0,\\
\mathrm{Var}(z_i)=&\sum_{i=1}^p\left(a_i^2\cdot1\right)=1.
\end{align*}

The expected squared distance of $z_i$ is:
$$\mathrm{E}[\mathrm{d}(z_i,0)^2]=\mathrm{E}[z_i^2]=1.$$

The expected squared distance for $x_0$ is:
$$\mathrm{E}[\mathrm{d}(x_0,0)^2]=\mathrm{E}[x_0^2]=p,$$
since $x_0^2 \sim \chi^2(p).$

# Ex. 2.5

## (a)
This decomposition of expected prediction error is a general law for any prediction, and should be memorized:
\begin{align*}
\mathrm{E}\left[\left(y_{0}-\hat{y}_{0}\right)^{2}\right] & =\mathrm{E}\left[\left(y_{0}-\mathrm{E}[\hat{y}_{0}]+\mathrm{E}[\hat{y}_{0}]-\hat{y}_{0}\right)^{2}\right]\\
 & =\mathrm{E}\left[\left(y_{0}-\mathrm{E}[\hat{y}_{0}]\right)^{2}\right]+\mathrm{E}\left[\left(\mathrm{E}[\hat{y}_{0}]-\hat{y}_{0}\right)^{2}\right]+0\\
 & =\mathrm{E}\left[\left(y_{0}-\mathrm{E}[y_{0}]+\mathrm{E}[y_{0}]-\mathrm{E}[\hat{y}_{0}]\right)^{2}\right]+\mathrm{Var}[\hat{y}_{0}]\\
 & =\mathrm{E}\left[\left(y_{0}-\mathrm{E}[y_{0}]\right)^{2}\right]+\mathrm{E}\left[\left(\mathrm{E}[y_{0}]-\mathrm{E}[\hat{y}_{0}]\right)^{2}\right]+0+\mathrm{Var}[\hat{y}_{0}]\\
 & =\mathrm{Var}[y_{0}]+\left(\mathrm{E}[y_{0}]-\mathrm{E}[\hat{y}_{0}]\right)^{2}+\mathrm{Var}[\hat{y}_{0}]\\
 & =\mathrm{Var}[y_{0}]+\mathrm{Bias}\left(\hat{y}_{0}\right)^{2}+\mathrm{Var}[\hat{y}_{0}]
\end{align*}
Following the notion in (2.4), where $\mathbf{X}$ is an $N\times p$ matrix with each row an input vector of the training set, and $\mathbf{y}$ is a $N$-vector of the output of the training set. The least square estimate of $\beta$ is given by:
$$\hat{\beta}=\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\mathbf{X}^T\mathbf{y}.$$
Then we have:
\begin{align*}
    y_0 - \hat{y}_0 =& x_0^T\beta+\varepsilon_0-x_0^T\hat{\beta} \\
    =& x_0^T\beta+\varepsilon_0
    -x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\mathbf{X}^T\mathbf{y} \\
    =& x_0^T\beta+\varepsilon_0 - x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}
    \mathbf{X}^T\left(\mathbf{X}\beta+\varepsilon\right) \\
    =& \varepsilon_0 - x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\mathbf{X}^T\varepsilon.
\end{align*}
Knowing that $\varepsilon_0$ and $\varepsilon$ are independent:
\begin{align*}
    \mathrm{EPE}(x_0)=&\mathrm{E}[(y_0 - \hat{y}_0)^2]\\
    =&\mathrm{E}\left[\left(\varepsilon_0 - x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}
    \mathbf{X}^T\varepsilon\right)^2\right] \\
    =&\mathrm{E}\left[\varepsilon_0^2\right] + \mathrm{E}\left[
    x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\mathbf{X}^T\varepsilon 
    \varepsilon^T\mathbf{X}\left(\mathbf{X}^T\mathbf{X}\right)^{-1}x_0\right]+0 \\
    =&\mathrm{E}\left[\varepsilon_0^2\right] + \mathrm{E}\left[
    x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\mathbf{X}^T\mathrm{E}\left[\varepsilon 
    \varepsilon^T\right]\mathbf{X}\left(\mathbf{X}^T\mathbf{X}\right)^{-1}x_0\right]+0 \\
    =&\sigma^2+\sigma^2\mathrm{E}\left[
    x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\mathbf{X}^T\cdot\mathbf{I}\cdot
    \mathbf{X}\left(\mathbf{X}^T\mathbf{X}\right)^{-1}x_0\right] \\
    =&\sigma^2+\sigma^2\mathrm{E}\left[x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}x_0\right].
\end{align*}
According to the decomposition of expected prediction error:
\begin{align*}
\mathrm{Var}[y_{0}] & =\sigma^{2}\\
\mathrm{Bias}\left(\hat{y}_{0}\right)^{2} & =0\\
\mathrm{Var}[\hat{y}_{0}] & =\sigma^{2}\mathrm{E}\left[x_{0}^{T}\left(\mathbf{X}^{T}\mathbf{X}\right)^{-1}x_{0}\right]
\end{align*}
In the fourth line of the above equations, I assumed the following:
$$\mathrm{E}[\mathbf{XYZ}]=\mathrm{E}[\mathbf{X}\cdot\mathrm{E}[\mathbf{Y}]\cdot\mathbf{Z}],$$
where $\mathbf{X}$,$\mathbf{Y}$ and $\mathbf{Z}$ are random matrices, and $\mathbf{Y}$ is independent of both $\mathbf{X}$ and $\mathbf{Z}$. Let's take a step back to first prove for the scalar case: 
$$\mathrm{E}[\mathbf{xyz}]=\mathrm{E}[\mathbf{x}\cdot\mathrm{E}[\mathbf{y}]\cdot\mathbf{z}].$$

By law of total expectation (or tower rule):
\begin{align*}
\mathrm{E}[\mathbf{xyz}] & =\mathrm{E}[\mathrm{E}[\mathbf{xyz}|\mathbf{y}]]\\
 & =\mathrm{E}[\mathrm{E}[\mathbf{y}|\mathbf{y}]\cdot\mathrm{E}[\mathbf{xz}|\mathbf{y}]]\\
 & =\mathrm{E}[\mathbf{y}\cdot\mathrm{E}[\mathbf{xz}|\mathbf{y}]]\\
 & =\mathrm{E}[\mathbf{y}\cdot\mathrm{E}[\mathbf{xz}]]\qquad\mbox{by independence of}\,\,\mathbf{y}\\
 & =\mathrm{E}[\mathbf{y}]\cdot\mathrm{E}[\mathbf{xz}]\\
 & =\mathrm{E}[\mathbf{x}\cdot\mathrm{E}[\mathbf{y}]\cdot\mathbf{z}]
\end{align*}

The matrix case is straight forward by definition of matrix mutiplication:
\begin{align*}
\mathrm{E}[\mathbf{XYZ}]_{i,j} & =\mathrm{E}\left[\sum_{k}\sum_{l}\mathbf{X}_{i,k}\mathbf{Y}_{k,l}\mathbf{Z}_{l,j}\right]\\
 & =\sum_{k}\sum_{l}\mathrm{E}[\mathbf{X}_{i,k}\mathbf{Y}_{k,l}\mathbf{Z}_{l,j}]\\
 & =\sum_{k}\sum_{l}\mathrm{E}\left[\mathbf{X}_{i,k}\mathrm{E}[\mathbf{Y}_{k,l}]\mathbf{Z}_{l,j}\right]\\
 & =\sum_{k}\sum_{l}\mathrm{E}\left[\mathbf{X}_{i,k}\mathrm{E}[\mathbf{Y}]_{k,l}\mathbf{Z}_{l,j}\right]\\
 & =\mathrm{E}\left[\sum_{k}\sum_{l}\mathbf{X}_{i,k}\mathrm{E}[\mathbf{Y}]_{k,l}\mathbf{Z}_{l,j}\right]\\
 & =\mathrm{E}[\mathbf{X}\cdot\mathrm{E}[\mathbf{Y}]\cdot\mathbf{Z}]_{i,j}
\end{align*}

## (b)

Assuming $\mathrm{E}[\mathbf{X}]=0$, by law of large number we have 
$$\frac{\mathbf{X}^T\mathbf{X}}{N}\rightarrow \mathrm{Cov}(\mathbf{x})\, 
\mathrm{when}\, N\rightarrow\infty.$$
Further more, 
$$\frac{\mathrm{Cov}(\mathbf{x})^{-1}\mathbf{X}^T\mathbf{X}}{N}\rightarrow \mathbf{I}\, 
\mathrm{when}\, N\rightarrow\infty.$$
$x_0$ has the same distribution as any row of $\mathbf{X}$:
$$\mathrm{E}[x_0x_0^T]=\mathrm{Cov}(x_0)=\mathrm{Cov}(\mathbf{x}).$$
Knowing that $x_0$ and $\mathbf{X}$ are independent, and that both $\mathrm{Cov}(\mathbf{x})$ and $\mathbf{X}^T\mathbf{X}$ are symmetric and invertible (for the latter, almost surely):
\begin{align*}
\mathrm{E}\left[x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}x_0\right] =& 
\mathrm{E}\left[\mathrm{Tr}\left(x_0x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\right)\right]\\
=&\mathrm{Tr}\left(\mathrm{E}\left[x_0x_0^T\right]\cdot
\mathrm{E}\left[\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\right]\right)\\
=&\mathrm{Tr}\left(\mathrm{Cov}(\mathbf{x})\cdot
\mathrm{E}\left[\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\right]\right)\\
=&\mathrm{Tr}\left(\mathrm{E}\left[\mathrm{Cov}(\mathbf{x})
\cdot\left(\mathbf{X}^T\mathbf{X}\right)^{-1}\right]\right)\\
=& \mathrm{Tr}\left(\mathrm{E}\left[\left(\mathrm{Cov}(\mathbf{x})^{-1}
\cdot\mathbf{X}^T\mathbf{X}\right)^{-1}\right]\right).\\
\end{align*}

The matrix inverse operator is a continuous operator on the space of $p\times p$ invertible matrices, as a result, we have:
$$\left(\mathrm{Cov}(\mathbf{x})^{-1}
\cdot\mathbf{X}^T\mathbf{X}\right)^{-1}\sim \left(N\cdot\mathbf{I}\right)^{-1}\,
\mathrm{when}\, N\rightarrow\infty.$$
One may notice that the matrix inverse operator is continuous but not uniformly continuous, so the above limit behavior holds true almost surely, since the uninvertible $p\times p$ matrices consists a negligible subset of the set of all $p\times p$ matrices. 

Finally we have:
\begin{align*}
\mathrm{Var}[\hat{y}_{0}]=
\sigma^2\mathrm{E}\left[x_0^T\left(\mathbf{X}^T\mathbf{X}\right)^{-1}x_0\right] \sim&
\sigma^2\mathrm{Tr}\left(\mathrm{E}\left[\left(N\cdot\mathbf{I}\right)^{-1}\right]\right)\\
=& \sigma^2N^{-1}\mathrm{Tr}(\mathbf{I})\\
=& \sigma^2p/N.
\end{align*}
This completes the derivation of (2.28). 

# Ex. 2.6

This problem asks us to prove that the simple least square estimator of $\theta$, equals some form of weighted least square estimator of $\theta$ when some values of $x_i$ are repeated. Suppose we have $N_i$ repeatitions of $x_i$, with outputs $y_{ij}$ where $j=1,\dots,N_i$. For input $x_i$ natrually we would guess that the weight for this data entry to be $N_i$, and the output to be the everage of all $N_j$ numbers of $y_{ij}$, let's denote it as $\bar{y_i}$. Now if we are lucky, the following can be proved:
$$\mathrm{argmin}_\theta\sum_{i=1}^N\sum_{j=1}^{N_i}(f_\theta(x_i)-y_{ij})^2
=\mathrm{argmin}_\theta\sum_{i=1}^NN_i(f_\theta(x_i)-\bar{y_i})^2$$
**Proof**
\begin{align*}
\mathrm{argmin}_\theta\sum_{i=1}^N\sum_{j=1}^{N_i}(f_\theta(x_i)-y_{ij})^2
=&\mathrm{argmin}_\theta\sum_{i=1}^N\sum_{j=1}^{N_i}
\left(f_\theta(x_i)^2-2y_{ij}f_\theta(x_i)+y_{ij}^2\right)\\
=&\mathrm{argmin}_\theta\sum_{i=1}^NN_i\left(f_\theta(x_i)^2-2f_\theta(x_i)\bar{y_i}
+\bar{y_i}^2-\bar{y_i}^2+\frac1{N_i}\sum_{j=1}^{N_i}y_{ij}^2\right)\\
=&\mathrm{argmin}_\theta\left(\sum_{i=1}^NN_i\left(f_\theta(x_i)-\bar{y_i}\right)^2
+\sum_{i=1}^N\left(\sum_{j=1}^{N_i}y_{ij}^2-N_i\bar{y_i}^2\right)\right)\\
=&\mathrm{argmin}_\theta\sum_{i=1}^NN_i\left(f_\theta(x_i)-\bar{y_i}\right)^2.
\end{align*}∎

The last line is based on the fact that $\sum_{i=1}^N\left(\sum_{j=1}^{N_i}y_{ij}^2-N_i\bar{y_i}^2\right)$ is not a function of $\theta$. 

# Ex. 2.7

Without loss of generality, let's assume $\mathbb{R}^p\overset{f}{\rightarrow}\mathbb{R}$. Namely, let $\mathbf{x}$ be the p dimensional random vector with $\{x_i | i=1,\dots,N\}$ as its $N$ realizations. 
## (a)
In the case of linear regression, $f(\mathbf{x})=\mathbf{x}^T\beta$. $\mathcal{X}$ will a $N\times p$ dimensional matrix with its $i$th row being $x_i^T$. The least square esimator of $\beta$ is:
$$\hat{\beta}=\left(\mathcal{X}^T\mathcal{X}\right)^{-1}\mathcal{X}^T\mathcal{Y}.$$
Therefore:
\begin{align*}
\hat{f}(x_0)=&x_0^T\hat{\beta}\\
=&x_0^T\left(\mathcal{X}^T\mathcal{X}\right)^{-1}\mathcal{X}^T\mathcal{Y}\\
=&\sum_{i=1}^Nl_i(x_0;\mathcal{X})y_i,
\end{align*}
where $l_i(x_0;\mathcal{X})$ is the $i$th element of the row vector $x_0^T\left(\mathcal{X}^T\mathcal{X}\right)^{-1}\mathcal{X}^T$, and $y_i$ is the $i$th element of the column vector of $\mathcal{Y}$.

In the case of $k$-nearest-neighbor, $l_i(x_0;\mathcal{X})$ is either $1/k$, if the $x_i$ is in the set of $k$-nearest-neighbor of $x_0$, or $0$, if otherwise.

In both cases $l_i(x_0;\mathcal{X})$ doesn't depend on $y_i$.
## (b)
This is my first serious study in statistics and was confused about the notation at first. With help from a friend I learned that $\mathbf{E}_{X,Y|Z}[\cdot]$ means the same thing as $\mathbf{E}[\cdot|Z]$ in probability. It seems that statisticians love to keep themselves reminded about what the random variables are, and I can imagine that for regression there can be a lot of them. Weirdly, (but with all due respet) the unconditinal epxectation $\mathbf{E}[\cdot]$ in probability becomes $\mathbf{E}_{X,Y}[\cdot]$. Since we are learning statistics here, let's follow the notation.
\begin{align*}
\mathbf{E}_{\mathcal{Y|X}}\left(f(x_0)-\hat{f}(x_0)\right)^2=&
\mathbf{E}_{\mathcal{Y|X}}\left(f(x_0) - \mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)
+\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0) -\hat{f}(x_0)\right)^2\\
=&\mathbf{E}_{\mathcal{Y|X}}\left(f(x_0)
-\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)\right)^2
+\mathbf{E}_{\mathcal{Y|X}}\left(\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)-\hat{f}(x_0)\right)^2\\
&+2\mathbf{E}_{\mathcal{Y|X}}\left[\left(f(x_0)-\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)\right)
\left(\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)-\hat{f}(x_0)\right)\right]
\end{align*}
In the last line, the first term is just $\left(f(x_0)-\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)\right)^2$, because it is not random given $\mathcal{X}$. The difference between the expection of an estimator and the true value is the bias of the estimator, so the first term is the square of the bias of $\hat{f}(x_0)$. The second term is the variance of $\hat{f}(x_0)$ by definition. Let's take a look at the third term. Since $f(x_{0})-\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_{0})$ is non-random, we have:
\begin{align*}
\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\left[\left(f(x_{0})-\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})\right)\left(\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})-\hat{f}(x_{0})\right)\right] & =\left(f(x_{0})-\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})\right)\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\left[\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})-\hat{f}(x_{0})\right]\\
 & =\left(f(x_{0})-\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})\right)\left(\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})-\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})\right)\\
 & =\left(f(x_{0})-\mathbf{E}_{\mathcal{Y}|\mathcal{X}}\hat{f}(x_{0})\right)\cdot0=0
\end{align*}

Finally, we have:
\begin{align*}
\mathbf{E}_{\mathcal{Y|X}}\left(f(x_0)-\hat{f}(x_0)\right)^2
=&\mathrm{Bias}(\hat{f}(x_{0})|\mathcal{X})^2+\mathrm{Var}(\hat{f}(x_{0})|\mathcal{X})\\
=&\left(f(x_0)-\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)\right)^2 
+ \mathbf{E}_{\mathcal{Y|X}}\left(\mathbf{E}_{\mathcal{Y|X}}\hat{f}(x_0)-\hat{f}(x_0)\right)^2
\end{align*}

## (c)
The derivation is the same for unconditional expected prediction error and the result is:
\begin{align*}
\mathbf{E}_{\mathcal{X},\mathcal{Y}}\left(f(x_{0})-\hat{f}(x_{0})\right)^{2} & =\mathrm{Bias}(\hat{f}(x_{0}))^{2}+\mathrm{Var}(\hat{f}(x_{0}))\\
 & =\left(\mathbf{E}_{\mathcal{X},\mathcal{Y}}\hat{f}(x_{0})-f(x_{0})\right)^{2}+\mathbf{E}_{\mathcal{X},\mathcal{Y}}\left(\hat{f}(x_{0})-\mathbf{E}_{\mathcal{X},\mathcal{Y}}\left[\hat{f}(x_{0})\right]\right)^{2}
\end{align*}

## (d)
I don't quite understand what this problem is asking for. I'm writing down the results for linear regression when the true model is indeed a linear model. 
For the linear regression, the bias is zero regardless of conditioning. For the variance conditional on $\mathcal{X}$, we have:
\begin{align*}
\mathrm{Var}[\hat{f}(x_{0})|\mathcal{X}] & =\mathrm{E}\left[x_{0}^{T}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}\mathcal{X}^{T}\varepsilon\varepsilon^{T}\mathcal{X}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}x_{0}|\mathcal{X}\right]\\
 & =x_{0}^{T}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}\mathcal{X}^{T}\mathrm{E}\left[\varepsilon\varepsilon^{T}|\mathcal{X}\right]\mathcal{X}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}x_{0}\\
 & =x_{0}^{T}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}\mathcal{X}^{T}\sigma^{2}\mathbf{I}\mathcal{X}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}x_{0}\\
 & =\sigma^{2}x_{0}^{T}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}\mathcal{X}^{T}\mathcal{X}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}x_{0}\\
 & =\sigma^{2}x_{0}^{T}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}x_{0}
\end{align*}
For the unconditionl variance, following Ex. 2.5 (a):
$$\mathrm{Var}[\hat{f}(x_{0})] =\sigma^{2}\mathrm{E}\left[x_{0}^{T}\left(\mathcal{X}^{T}\mathcal{X}\right)^{-1}x_{0}\right]$$

# Ex. 2.8
The data files are download from the book website: www-stat.stanford.edu/ElemStatLearn, and stored at 'data/Chapter02'. Let's first load the data as pandas data frames.

In [1]:
import pandas as pd
import numpy as np
from os import getcwd
fname_train = getcwd()+'/data/Chapter02/zip.train'
fname_train2 = getcwd()+'/data/Chapter02/train.2'
fname_train3 = getcwd()+'/data/Chapter02/train.3'
fname_test = getcwd()+'/data/Chapter02/zip.test'
colnames = ['Digit']+['Grayvalue'+str(i) for i in range(0,256)]

train_dat=pd.read_table(fname_train,sep=' ',header=None,names=colnames,index_col=False)
train2_dat=pd.read_table(fname_train2,sep=',',header=None,names=colnames[1:],index_col=False)
train3_dat=pd.read_table(fname_train3,sep=',',header=None,names=colnames[1:],index_col=False)
test_dat=pd.read_table(fname_test,sep=' ',header=None,names=colnames,index_col=False)

num_train = train_dat.shape[0]
num_train2 = train2_dat.shape[0]
num_train3 = train3_dat.shape[0]
num_test = test_dat.shape[0]
# add exogenenous variable to fit the intercept
train_dat.insert(1,'intercept',pd.Series(np.ones(num_train),index=train_dat.index))
train2_dat.insert(0,'intercept',pd.Series(np.ones(num_train2),index=train2_dat.index))
train3_dat.insert(0,'intercept',pd.Series(np.ones(num_train3),index=train3_dat.index))
test_dat.insert(1,'intercept',pd.Series(np.ones(num_test),index=test_dat.index))

Let's firt try the ordinary linear least square with no intercept. In this case, a picture with gray scale value of 0 everywhere means 0. 

In [2]:
from statsmodels.api import OLS

result_ols = OLS(endog=train_dat['Digit'], exog=train_dat[colnames[1:]]).fit()
result_ols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Digit   R-squared:                       0.872
Model:                            OLS   Adj. R-squared:                  0.868
Method:                 Least Squares   F-statistic:                     187.6
Date:                Sat, 26 Mar 2016   Prob (F-statistic):               0.00
Time:                        17:22:13   Log-Likelihood:                -14463.
No. Observations:                7291   AIC:                         2.944e+04
Df Residuals:                    7035   BIC:                         3.120e+04
Df Model:                         256                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [95.0% Conf. Int.]
--------------------------------------------------------------------------------
Grayvalue0      -2.2494      0.632     -3.560      0.000        -3.488    -1.011
Grayvalue1       0.1097      0.319      0.344      0.731        -0.516     0.735
Grayvalue2      -0.1743      0.214     -0.813      0.416        -0.594     0.246
Grayvalue3      -0.0129      0.147     -0.087      0.930        -0.302     0.276
Grayvalue4      -0.2128      0.115     -1.848      0.065        -0.438     0.013
Grayvalue5       0.0764      0.094      0.814      0.416        -0.108     0.260
Grayvalue6      -0.1731      0.076     -2.280      0.023        -0.322    -0.024
Grayvalue7      -0.5274      0.065     -8.062      0.000        -0.656    -0.399
Grayvalue8       0.1279      0.063      2.038      0.042         0.005     0.251
Grayvalue9       0.5461      0.069      7.860      0.000         0.410     0.682
Grayvalue10      0.3896      0.077      5.042      0.000         0.238     0.541
Grayvalue11      0.0134      0.089      0.151      0.880        -0.160     0.187
Grayvalue12      0.1941      0.109      1.783      0.075        -0.019     0.408
Grayvalue13      0.4146      0.149      2.779      0.005         0.122     0.707
Grayvalue14      0.0300      0.222      0.135      0.893        -0.405     0.465
Grayvalue15      0.7707      0.429      1.797      0.072        -0.070     1.611
Grayvalue16     -0.4275      0.397     -1.076      0.282        -1.206     0.351
Grayvalue17     -0.5740      0.227     -2.532      0.011        -1.018    -0.130
Grayvalue18     -0.2431      0.147     -1.657      0.098        -0.531     0.044
Grayvalue19     -0.2627      0.109     -2.415      0.016        -0.476    -0.049
Grayvalue20     -0.1723      0.091     -1.891      0.059        -0.351     0.006
Grayvalue21     -0.0794      0.084     -0.940      0.347        -0.245     0.086
Grayvalue22      0.0149      0.081      0.183      0.855        -0.144     0.174
Grayvalue23     -0.0059      0.079     -0.075      0.940        -0.161     0.149
Grayvalue24     -0.0326      0.077     -0.424      0.671        -0.183     0.118
Grayvalue25      0.2445      0.073      3.368      0.001         0.102     0.387
Grayvalue26      0.1705      0.072      2.363      0.018         0.029     0.312
Grayvalue27      0.1748      0.081      2.148      0.032         0.015     0.334
Grayvalue28      0.2745      0.101      2.707      0.007         0.076     0.473
Grayvalue29     -0.0626      0.134     -0.467      0.641        -0.325     0.200
Grayvalue30      0.5375      0.193      2.787      0.005         0.159     0.916
Grayvalue31     -0.0257      0.304     -0.085      0.933        -0.621     0.570
Grayvalue32     -0.2479      0.282     -0.881      0.379        -0.800     0.304
Grayvalue33      0.2480      0.183      1.353      0.176        -0.111     0.607
Grayvalue34     -0.2158      0.123     -1.751      0.080        -0.457     0.026
Grayv

Next we examine the OLS model fitting results by testing on both the training set and test set. We round the prediction to the nearest integer, with a floor of 0 and a cap of 9. The we count the percetage of predictions that is correct. We achieve roughly 20~30% success rate for digit recognition by using OLS with no intercept.

In [3]:
# prediction using the entire test set
pred_test = result_ols.predict(test_dat[colnames[1:]])
pred_test=np.round(pred_test)
pred_test[pred_test<0]=0
pred_test[pred_test>9]=9
print("Correct rate on test set:")
print(sum(pred_test==test_dat['Digit'])/num_test)
# prediction using the entire training set
pred_train = result_ols.predict(train_dat[colnames[1:]])
pred_train=np.round(pred_train)
pred_train[pred_train<0]=0
pred_train[pred_train>9]=9
print("Correct rate on entire train set:")
print(sum(pred_train==train_dat['Digit'])/num_train)
# prediction using the training set with the digit being 2
pred_train2 = result_ols.predict(train2_dat[colnames[1:]])
pred_train2=np.round(pred_train2)
pred_train2[pred_train2<0]=0
pred_train2[pred_train2>9]=9
print("Correct rate on training set with the digit being 2:")
print(sum(pred_train2==2)/num_train2)
# prediction using the training set with the digit being 3
pred_train3 = result_ols.predict(train3_dat[colnames[1:]])
pred_train3=np.round(pred_train2)
pred_train3[pred_train3<0]=0
pred_train3[pred_train3>9]=9
print("Correct rate on training set with the digit being 3:")
print(sum(pred_train3==3)/num_train3)

Correct rate on test set:
0.254608868959
Correct rate on entire train set:
0.269784666027
Correct rate on training set with the digit being 2:
0.243502051984
Correct rate on training set with the digit being 3:
0.306990881459


Next we try the OLS with intercept. In this case zero grey scale value every where represents a digit specified by the intercept rounded up to the nearest integer. 

In [4]:
from statsmodels.api import OLS

result_ols_intercept = OLS(endog=train_dat['Digit'], exog=train_dat[['intercept']+colnames[1:]]).fit()
result_ols_intercept.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Digit   R-squared:                       0.655
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     52.25
Date:                Sat, 26 Mar 2016   Prob (F-statistic):               0.00
Time:                        17:22:14   Log-Likelihood:                -14463.
No. Observations:                7291   AIC:                         2.944e+04
Df Residuals:                    7034   BIC:                         3.121e+04
Df Model:                         256                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [95.0% Conf. Int.]
--------------------------------------------------------------------------------
intercept        0.6094      1.000      0.610      0.542        -1.350     2.569
Grayvalue0      -2.1020      0.677     -3.107      0.002        -3.428    -0.776
Grayvalue1       0.0912      0.320      0.285      0.776        -0.537     0.719
Grayvalue2      -0.1640      0.215     -0.763      0.446        -0.585     0.257
Grayvalue3      -0.0123      0.147     -0.084      0.933        -0.301     0.277
Grayvalue4      -0.2103      0.115     -1.825      0.068        -0.436     0.016
Grayvalue5       0.0771      0.094      0.822      0.411        -0.107     0.261
Grayvalue6      -0.1719      0.076     -2.263      0.024        -0.321    -0.023
Grayvalue7      -0.5271      0.065     -8.056      0.000        -0.655    -0.399
Grayvalue8       0.1285      0.063      2.047      0.041         0.005     0.252
Grayvalue9       0.5471      0.070      7.872      0.000         0.411     0.683
Grayvalue10      0.3904      0.077      5.051      0.000         0.239     0.542
Grayvalue11      0.0151      0.089      0.170      0.865        -0.159     0.189
Grayvalue12      0.1951      0.109      1.791      0.073        -0.018     0.409
Grayvalue13      0.4200      0.149      2.810      0.005         0.127     0.713
Grayvalue14      0.0255      0.222      0.115      0.909        -0.410     0.461
Grayvalue15      0.8472      0.447      1.896      0.058        -0.029     1.723
Grayvalue16     -0.4202      0.398     -1.057      0.291        -1.200     0.359
Grayvalue17     -0.5871      0.228     -2.578      0.010        -1.034    -0.141
Grayvalue18     -0.2423      0.147     -1.651      0.099        -0.530     0.045
Grayvalue19     -0.2652      0.109     -2.436      0.015        -0.479    -0.052
Grayvalue20     -0.1715      0.091     -1.883      0.060        -0.350     0.007
Grayvalue21     -0.0798      0.084     -0.946      0.344        -0.245     0.086
Grayvalue22      0.0150      0.081      0.185      0.854        -0.144     0.174
Grayvalue23     -0.0061      0.079     -0.077      0.939        -0.161     0.149
Grayvalue24     -0.0328      0.077     -0.428      0.669        -0.183     0.118
Grayvalue25      0.2435      0.073      3.353      0.001         0.101     0.386
Grayvalue26      0.1702      0.072      2.358      0.018         0.029     0.312
Grayvalue27      0.1734      0.081      2.129      0.033         0.014     0.333
Grayvalue28      0.2722      0.101      2.683      0.007         0.073     0.471
Grayvalue29     -0.0653      0.134     -0.487      0.626        -0.328     0.198
Grayvalue30      0.5346      0.193      2.771      0.006         0.156     0.913
Grayvalue31     -0.0451      0.305     -0.148      0.883        -0.644     0.554
Grayvalue32     -0.2229      0.285     -0.783      0.434        -0.781     0.335
Grayvalue33      0.2403      0.184      1.308      0.191        -0.120     0.601
Grayv

The flollowing result showed no much difference.

In [5]:
# prediction using the entire test set
pred_test = result_ols_intercept.predict(test_dat[['intercept']+colnames[1:]])
pred_test=np.round(pred_test)
pred_test[pred_test<0]=0
pred_test[pred_test>9]=9
print("Correct rate on test set:")
print(sum(pred_test==test_dat['Digit'])/num_test)
# prediction using the entire training set
pred_train = result_ols_intercept.predict(train_dat[['intercept']+colnames[1:]])
pred_train=np.round(pred_train)
pred_train[pred_train<0]=0
pred_train[pred_train>9]=9
print("Correct rate on entire train set:")
print(sum(pred_train==train_dat['Digit'])/num_train)
# prediction using the training set with the digit being 2
pred_train2 = result_ols_intercept.predict(train2_dat[['intercept']+colnames[1:]])
pred_train2=np.round(pred_train2)
pred_train2[pred_train2<0]=0
pred_train2[pred_train2>9]=9
print("Correct rate on training set with the digit being 2:")
print(sum(pred_train2==2)/num_train2)
# prediction using the training set with the digit being 3
pred_train3 = result_ols_intercept.predict(train3_dat[['intercept']+colnames[1:]])
pred_train3=np.round(pred_train2)
pred_train3[pred_train3<0]=0
pred_train3[pred_train3>9]=9
print("Correct rate on training set with the digit being 3:")
print(sum(pred_train3==3)/num_train3)

Correct rate on test set:
0.254608868959
Correct rate on entire train set:
0.268824578247
Correct rate on training set with the digit being 2:
0.236662106703
Correct rate on training set with the digit being 3:
0.316109422492


Now let's implement $k$-nearest-neighbor by brute force. We use a max heap to store $k$ nodes nearest to the test point and update it iteratively with all the training points. The time complexity should be $O(MN)$, where $M$ is the size of the training set and $N$ is the test set. Another simple algorithm is k-d-tree, and cannot be used here. k-d-tree performs poorly and is almost the same as brute force as the dimmension here is 256, and $2^{256} >> 7000$.

On my machine, one test will take about 7 seconds. Let's test for only 100 points and print the success rate.

In [6]:
from time import time
import heapq
gray_scale_col = colnames[1:]
def knn(k, p_test, training_set):
    def distance(p_test, p_train):
        return np.square(p_test[gray_scale_col].values-p_train[gray_scale_col].values).sum()
    minhp = [] # python only have min heap, use the negative of distance as priority value
    train_len = training_set.shape[0]
    if train_len <= k:
        pred = np.mean(training_set['Digit'])
    else:
        for i in range(k):
            heapq.heappush(minhp, (-distance(p_test, training_set.iloc[i]), i))
        for i in range(k, train_len):
            priority = -distance(p_test, training_set.iloc[i])
            if priority > minhp[0][0]:
                heapq.heapreplace(minhp, (priority, i))
        row_indices = [entry[1] for entry in minhp]
        pred = np.mean(training_set['Digit'][row_indices].values)
    pred=np.round(pred)
    if pred > 9:
        return 9
    elif pred <0:
        return 0
    else:
        return int(pred)
tic = time()
Ntests = 100
count = 0.0
for i in range(Ntests):
    pred = knn(15, test_dat.iloc[i], train_dat)
    if pred == int(test_dat['Digit'][i]):
        count+=1.0
print("Time cost in minutes:")
print((time()-tic)/60.0)
print("Success rate:")
print(count/Ntests)

Time cost in minutes:
12.342083319028218
Success rate:
0.84


# Ex. 2.9

This is a natrual result of least square estimator being Best Linear Unbiased Estimator (BLUE). Let's first try if we can explicitly calculate $\mathrm{E}[R_{tr}(\hat{\beta})]$ and $\mathrm{E}[R_{te}(\hat{\beta})]$ and compare their values. Finally we will prove the BLUE property of least square estimator and use it to derive the desired result.

Let's use $\mathbf{X}$ to denote the $N\times p$ matrix with its $i$th row being $x_i$ in the training set and $\mathbf{y}$ being the $N\times 1$ vector with its $i$th element being $y_i$ in the training set. Let's denote the true value of $\beta$ as $\beta_{true}$, then the true model is as follows:
$$\mathbf{y}=\mathbf{X}\beta_{\mbox{true}}+\mathbf{\varepsilon},$$
where $\mathbf{\varepsilon}$ is the random error vector with zero mean and covariance of $\sigma^2\mathbf{I}$.

The least square estimator of $\beta_{true}$ is:
$$\hat{\beta}=(\mathbf{X}^{T}\mathbf{X})^{-1}\mathbf{X}^{T}\mathbf{y}.$$

Next we can derive $\mathrm{E}[R_{tr}(\hat{\beta})]$:
\begin{align*}
\mathrm{E}[R_{tr}(\hat{\beta})] & =\mathrm{E}\left[\frac{1}{N}\sum_{i=1}^{N}(y_{i}-x_{i}^{T}\hat{\beta})\right]\\
 & =\frac{1}{N}\mathrm{E}\left[\sum_{i=1}^{N}(y_{i}-x_{i}^{T}\hat{\beta})\right]\\
 & =\frac{1}{N}\mathrm{E}\left[\left(\mathbf{X}\hat{\beta}-\mathbf{y}\right)^{T}\left(\mathbf{X}\hat{\beta}-\mathbf{y}\right)\right]\\
 & =\frac{1}{N}\mathrm{E}\left[\left(\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\mathbf{X}^{T}\mathbf{\varepsilon}-\mathbf{\varepsilon}\right)^{T}\left(\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\mathbf{X}^{T}\mathbf{\varepsilon}-\mathbf{\varepsilon}\right)\right]\\
 & =\frac{1}{N}\mathrm{E}\left[\mathbf{\varepsilon}^{T}\mathbf{\varepsilon}-\mathbf{\varepsilon}^{T}\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\mathbf{X}^{T}\mathbf{\varepsilon}\right]\\
 & =\frac{1}{N}\mathrm{E}\left[\mathbf{\varepsilon}^{T}\mathbf{\varepsilon}\right]-\frac{1}{N}\mathrm{E}\left[\mathbf{\varepsilon}^{T}\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\mathbf{X}^{T}\mathbf{\varepsilon}\right]\\
 & =\sigma^{2}-\frac{1}{N}\mathrm{E}\left[\mathbf{\varepsilon}^{T}\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\mathbf{X}^{T}\mathbf{\varepsilon}\right]\\
 & =\sigma^{2}-\frac{1}{N}\mathrm{E}\left[\mathrm{Tr}\left(\left(\mathbf{X}^{T}\mathbf{\varepsilon}\mathbf{\varepsilon}^{T}\mathbf{X}\right)(\mathbf{X}^{T}\mathbf{X})^{-1}\right)\right]\\
 & =\sigma^{2}-\frac{1}{N}\mathrm{E}\left[\mathrm{Tr}\left(\mathbf{X}^{T}\mathrm{E}\left[\mathbf{\varepsilon}\mathbf{\varepsilon}^{T}\right]\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\right)\right]\quad\mathbf{\varepsilon}\mbox{ being independent of }\mathbf{X}\\
 & =\sigma^{2}-\frac{\sigma^{2}}{N}\mathrm{E}\left[\mathrm{Tr}\left(\mathbf{X}^{T}\cdot\mathbf{I}\cdot\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\right)\right]\\
 & =\sigma^{2}-\frac{\sigma^{2}}{N}\mathrm{E}\left[\mathrm{Tr}\left(\mathbf{X}^{T}\mathbf{X}(\mathbf{X}^{T}\mathbf{X})^{-1}\right)\right]\\
 & =\sigma^{2}-\frac{\sigma^{2}}{N}\mathrm{E}\left[\mathrm{Tr}(\mathbf{I})\right]\\
 & =\sigma^{2}\left(1-\frac{p}{N}\right)
\end{align*}

For $\mathrm{E}[R_{te}(\hat{\beta})]$ we know that $\mathrm{E}[R_{te}(\hat{\beta})]=\mathrm{E}\left[\left(\tilde{y}_{1}-\tilde{x_{1}}\hat{\beta}\right)^{2}\right]$, due to the symmetry of each test point. Using the results in Ex. 2.5 (a), we know that:
$$\mathrm{E}\left[\left(\tilde{y}_{1}-\tilde{x_{1}}\hat{\beta}\right)^{2}\right] =\sigma^{2}\left(1+\mathrm{E}\left[\tilde{x}_{1}^{T}\left(\mathbf{X}^{T}\mathbf{X}\right)^{-1}\tilde{x}_{1}\right]\right)
$$
When $N>p$ we know that the matrix $\mathbf{X}^{T}\mathbf{X}$ is almost surely positive definte. Then its inverse is also almost surely positive defnite, so $\tilde{x}_{1}^{T}\left(\mathbf{X}^{T}\mathbf{X}\right)^{-1}\tilde{x}_{1}$ is positive almost surely. In fact we know from Ex. 2.5 (b), that it will converge to $p/N$ as $N$ gets large. Therefore, we come to the following relation:
\begin{align*}
\mathrm{E}[R_{te}(\hat{\beta})] & =\sigma^{2}\left(1+\mathrm{E}\left[\tilde{x}_{1}^{T}\left(\mathbf{X}^{T}\mathbf{X}\right)^{-1}\tilde{x}_{1}\right]\right)\\
 & >\sigma^{2}\\
 & >\sigma^{2}\left(1-\frac{p}{N}\right)\\
 & =\mathrm{E}[R_{tr}(\hat{\beta})]
\end{align*}
To complete the proof, let's quickly prove that the inverse of a postive definite matrix is also positive definite. Suppose matrix $A$ is positive definite so invertible, then $\forall x\ne0,\,\exists y\ne0,\,s.t.\,Ay=x \mbox{ and } x^{T}A^{-1}x=\left(y^{T}A^{T}\right)A^{-1}Ay=y^{T}A^{T}y=y^{T}Ay>0$.